crates/proto/src/tcp/tcp_stream.rs

```diff
@@ -161,25 +161,22 @@ impl<S: DnsTcpStream> TcpStream<S> {
         timeout: Duration,
         outbound_messages: StreamReceiver,
     ) -> Result<Self, io::Error> {
-        S::Time::timeout(timeout, future)
-            .map(move |tcp_stream: Result<Result<S, io::Error>, _>| {
-                tcp_stream
-                    .and_then(|tcp_stream| tcp_stream)
-                    .map(|tcp_stream| {
-                        debug!("TCP connection established to: {}", name_server);
-                        Self {
-                            socket: tcp_stream,
-                            outbound_messages,
-                            send_state: None,
-                            read_state: ReadTcpState::LenBytes {
-                                pos: 0,
-                                bytes: [0u8; 2],
-                            },
-                            peer_addr: name_server,
-                        }
-                    })
+        let tcp_stream = S::Time::timeout(timeout, future).await;
+        tcp_stream
+            .and_then(|tcp_stream| tcp_stream)
+            .map(|tcp_stream| {
+                debug!("TCP connection established to: {}", name_server);
+                Self {
+                    socket: tcp_stream,
+                    outbound_messages,
+                    send_state: None,
+                    read_state: ReadTcpState::LenBytes {
+                        pos: 0,
+                        bytes: [0u8; 2],
+                    },
+                    peer_addr: name_server,
+                }
             })
```

Comment:
What about using `??` here to remove more combinators and inlining the single-use binding. Is something like this viable?

```rust
    async fn connect_with_future<F: Future<Output = Result<S, io::Error>> + Send + 'static>(
        future: F,
        name_server: SocketAddr,
        timeout: Duration,
        outbound_messages: StreamReceiver,
    ) -> Result<Self, io::Error> {
        debug!("TCP connection established to: {}", name_server);
        Ok(Self {
            socket: S::Time::timeout(timeout, future).await??,
            outbound_messages,
            send_state: None,
            read_state: ReadTcpState::LenBytes {
                pos: 0,
                bytes: [0u8; 2],
            },
            peer_addr: name_server,
        })
    }
```

Reply:
Oops, edited my above suggestion. We don't want to inline `socket` into `Self` like I suggested at first because it needs to happen up front so the `debug!()` happens at the right time/place.